### ISSUE TITLE
Clarification on security/privacy

### ISSUE THREAD

**Issue by JimmyBoh** [mafintosh/hypercore#96]
My apologies if there is a better place to ask this!

In the Dat paper it says that there is a tradeoff with privacy when using discovery transports like the BitTorrent DHT. What does this actually mean? Does it mean that you can track which peers are asking for the same public key, but can't tell/access that data? How does the public key vs discovery key vs secret key fit into this? Is the public key the "network id" and the secret key the "password" to access that network?

I'm interested in using Hypercore as a distributed log for build times, exceptions, and any other metrics when working on project as a team. Data would only be shared among the team, integrated into things like webpack or gulp. I'd like to keep it as secure as possible, as filenames and stacktraces should be kept internal. Is this a good use case?

Thanks!

**Comment by joehand** [comment]
Hello!

Your use case sounds perfect, we've also started to integrate some of our logging and build tools with hypercore using [hyperpipe](https://github.com/mafintosh/hyperpipe) (apologies: docs need a bit of updating there) and [hypertail](https://github.com/yoshuawuyts/hypertail). However, I think we still need to update hypertail/hypername to be compatible with the newest hypercore release.

> Is the public key the "network id" and the secret key the "password" to access that network?

One of the key elements of Dat privacy is that the public key is never used in any discovery network, only the discovery key is (the hash of the public key). So in a sense, the *discovery key* is the "network id" and the *public key* is the "password".

>  Does it mean that you can track which peers are asking for the same public key, but can't tell/access that data? How does the public key vs discovery key vs secret key fit into this? 

You can track which peers are asking for the same discovery key. Once a user has the public key, they can access the data if they can connect to other users sharing that key.

The use of *discovery keys* is important in networks like DHT:

> For example if a Dat client decides to use the BitTorrent DHT to discover peers ... then because of the privacy design of the BitTorrent DHT it becomes public knowledge what key that client is searching for.

BitTorrent DHT is pretty "chatty" by design. Once you start searching for a *discovery key*, it becomes known knowledge. There are a some servers that may falsely respond to a request and try to connect. But unless they have the *public key*, they will not be able to connect. However, your IP address may become known in this process.

## tldr

As long as the public key isn't shared outside of your team, the content will be secure (though the IP addresses + discovery key may become known). In order of easy + less secure to harder + more secure, you can take a few steps further (mostly in discovery-swarm):

1. Disable bittorrent DHT discovery (using only DNS discovery)
2. Whitelist IP addresses
3. Run your own DNS discovery server
4. Encrypt contents before adding to hypercore (content is automatically encrypted in transit but this would also require decrypting after arrival).

We have a bit more on [security & privacy](https://docs.datproject.org/faq#security-and-privacy) as well, may answer any more lingering questions you had. Your questions will probably be good to include there are well, so I may copy some to a new issue in our docs.

**Comment by mafintosh** [comment]
What @joehand said. Also wanted to note that hypercore is network/discovery agnostic. It just wants you to pipe instances together. How you do that is up to you, so you can compose a network stack that fits your privacy requirements and replicate hypercores over that.

**Comment by mafintosh** [comment]
Feel free to re-open if you have more questions :)

**Comment by JimmyBoh** [comment]
Wow, thanks @joehand for the great explanation! And that makes sense @mafintosh, especially since the paper literally suggests carrier pigeons. 😄 

Thanks again gentlemen!
